functions/codepipeline-wait-handler/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/CodePipelineWaitHandler.java

```diff
@@ -0,0 +1,145 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public class CodePipelineWaitHandler implements RequestHandler<Map<String, Object>, Object> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CodePipelineWaitHandler.class);
+    private final CodePipelineClient codepipeline;
+
+    public CodePipelineWaitHandler() {
+        LOGGER.info("Version Info: {}", Utils.version(this.getClass()));
+        this.codepipeline = Utils.sdkClient(CodePipelineClient.builder(), CodePipelineClient.SERVICE_NAME);
+    }
+
+    @Override
+    public Object handleRequest(Map<String, Object> event, Context context) {
+        //logRequestEvent(event);
+
+        Map<String, Object> job = (Map<String, Object>) event.get("CodePipeline.job");
+        String jobId = (String) job.get("id");
+
+        LOGGER.info("Processing CloudFormation wait handle for CodePipeline job {}", jobId);
+        Map<String, Object> jobData = (Map<String, Object>) job.get("data");
+        Map<String, Object> actionConfig = (Map<String, Object>) jobData.get("actionConfiguration");
+        Map<String, String> configuration = (Map<String, String>) actionConfig.get("configuration");
+        String userParameters = configuration.get("UserParameters");
+        Map<String, Object> params = Utils.fromJson(userParameters, HashMap.class);
+        if (null == params) {
+            throw new RuntimeException("UserParameters can't be parsed as JSON");
+        }
+
+        try {
+            // Pre signed S3 URL
+            String waitHandle = (String) params.get("waitHandle");
+
+            // Since there's not built-in way to have dynamic CodePipeline stage actions,
+            // and because this pipeline could be triggered outside of CloudFormation, the
+            // wait condition handle mey be irrelevant. Skip it if it's out of date.
+            if (signalCloudFormation(waitHandle)) {
+                LOGGER.info("Signaling CloudFormation wait condition handle");
+
+                // Passed back to CloudFormation as !GetAtt WaitConditionResource.Data
+                String data = "";
+                // If this Lambda is invoked, the proceeding CodePipeline stage action succeeded
+                boolean success = true;
+                String failureReason = "";
+
+                // Signal the CloudFormation stack that it can continue
+                CloudFormationResponse.signal(waitHandle, success, jobId, data, failureReason);
+            } else {
+                LOGGER.info("Skipping CloudFormation signal for expired wait condition handle");
+            }
+
+            // Tell CodePipeline that it can continue
+            codepipeline.putJobSuccessResult(request -> request.jobId(jobId));
+        } catch (Exception e) {
+            LOGGER.error(Utils.getFullStackTrace(e));
+            failJob(jobId, "Error ", context);
+            throw e;
+        }
```

Comment:
It means if any processing we do as part of this stage's action blows up, we have to tell CodePipeline that the job failed. Since this is a Lambda invoke action we're responsible for telling CodePipeline whether to continue with the stage or not.